File: pipeline/reimpl/cocrdupc.py
"""
Reimplementation of COCRDUPC — CardDemo Credit Card Update Screen.

CICS online program. Reads an existing credit card for display,
then allows updating specific fields (embossed name, expiration date,
active status, CVV).

Navigation:
  ENTER  → look up card by number
  PF3    → return to calling program
  PF5    → save changes
  PF12   → return to card list without saving
"""

from __future__ import annotations
from dataclasses import dataclass
from typing import Optional

from .carddemo_data import (
    CarddemoCommarea, ScreenHeader, CardRecord,
    CCDA_MSG_INVALID_KEY, DFHENTER, DFHPF3, DFHPF5, DFHPF12,
)


WS_PGMNAME = "COCRDUPC"
WS_TRANID  = "CCUP"

DFHRESP_NORMAL = 0
DFHRESP_NOTFND = 13


@dataclass
class CardUpdateInput:
    card_num: str = ""
    embossed_name: str = ""
    expiration_date: str = ""
    active_status: str = ""
    cvv_cd: str = ""


@dataclass
class CardUpdateResult:
    card_found: Optional[CardRecord] = None
    commarea: Optional[CarddemoCommarea] = None
    screen: Optional[ScreenHeader] = None
    message: str = ""
    error: bool = False
    success: bool = False
    xctl_program: str = ""
    return_to_prev: bool = False


class CardRepository:
    def __init__(self, cards: dict[str, CardRecord]):
        self._cards = cards

    def find(self, card_num: str) -> Optional[CardRecord]:
        return self._cards.get(card_num.strip())

    def rewrite(self, card: CardRecord) -> int:
        key = card.card_num.strip()
        if key not in self._cards:
            return DFHRESP_NOTFND
        self._cards[key] = card
        return DFHRESP_NORMAL


def process_card_update(
    eibcalen: int,
    eibaid: str,
    commarea: CarddemoCommarea,
    card_input: CardUpdateInput,
    card_repo: CardRepository,
    preloaded_card_num: str = "",
) -> CardUpdateResult:
    """Process card update screen — mirrors COCRDUPC PROCEDURE DIVISION."""
    result = CardUpdateResult()

    if eibcalen == 0:
        result.return_to_prev = True
        result.xctl_program = "COSGN00C"
        return result

    result.commarea = commarea

    if commarea.cdemo_pgm_context == 0:
        commarea.cdemo_pgm_context = 1
        if preloaded_card_num:
            card = card_repo.find(preloaded_card_num)
            result.card_found = card
            if card is None:
                result.error = True
                result.message = f"Card {preloaded_card_num.strip()} not found"
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.commarea = commarea
        return result

    if eibaid in (DFHENTER, "ENTER"):
        return _lookup_card(card_input.card_num, card_repo, commarea, result)
    elif eibaid in (DFHPF3, "PF3"):
        back = commarea.cdemo_from_program or "COMEN01C"
        result.xctl_program = back
        result.return_to_prev = True
        result.commarea = commarea
        return result
    elif eibaid in (DFHPF5, "PF5"):
        return _save_card(card_input, card_repo, commarea, result)
    elif eibaid in (DFHPF12, "PF12"):
        result.xctl_program = "COCRDLIC"
        result.return_to_prev = True
        commarea.cdemo_to_program = "COCRDLIC"
        result.commarea = commarea
        return result
    else:
        result.error = True
        result.message = CCDA_MSG_INVALID_KEY
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result


def _lookup_card(card_num, repo, commarea, result):
    if not card_num or card_num.strip() == "":
        result.error = True
        result.message = "Card number can NOT be empty..."
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    card = repo.find(card_num)
    if card is None:
        result.error = True
        result.message = f"Card {card_num.strip()} not found"
    else:
        result.card_found = card
        result.message = "Press PF5 to save changes"

    result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
    result.screen.errmsg = result.message
    result.commarea = commarea
    return result


def _save_card(inp, repo, commarea, result):
    if not inp.card_num or inp.card_num.strip() == "":
        result.error = True
        result.message = "Card number can NOT be empty..."
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    card = repo.find(inp.card_num)
    if card is None:
        result.error = True
        result.message = f"Card {inp.card_num.strip()} not found"
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    # Apply updates
    if inp.embossed_name.strip():
        card.card_embossed_name = inp.embossed_name[:50]
    if inp.expiration_date.strip():
        card.card_expiration_date = inp.expiration_date[:10]
    if inp.active_status.strip():
        card.card_active_status = inp.active_status.strip()[0]
    if inp.cvv_cd.strip():
        try:
            card.card_cvv_cd = int(inp.cvv_cd.strip())
        except ValueError:
            result.error = True
            result.message = "CVV must be numeric"
            result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
            result.screen.errmsg = result.message
            result.commarea = commarea
            return result

    resp = repo.rewrite(card)
    if resp == DFHRESP_NORMAL:
        result.success = True
        result.card_found = card
        result.message = f"Card {card.card_num.strip()} has been updated ..."
    else:
        result.error = True
        result.message = "Unable to update card..."

    result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
    result.screen.errmsg = result.message
    result.commarea = commarea
    return result


# ── Differential harness runner adapter ──────────────────────────────────────

_SEED_CARDS: dict[str, CardRecord] = {
    "4111111111110001": CardRecord(card_num="4111111111110001", card_acct_id=10000001, card_cvv_cd=123,
                                   card_embossed_name="JANE USER", card_expiration_date="2028-12-31",
                                   card_active_status="Y"),
    "4111111111110002": CardRecord(card_num="4111111111110002", card_acct_id=10000001, card_cvv_cd=456,
                                   card_embossed_name="BOB SMITH", card_expiration_date="2027-06-30",
                                   card_active_status="Y"),
}


def run_vector(inputs: dict) -> dict:
    """Canonical runner entry point for the differential harness.

    SCENARIO selects a hardcoded test path:
      FIRST_ENTRY    — context=0, preloaded card → display card
      LOOKUP_CARD    — ENTER with card number → find card
      UPDATE_CARD    — PF5 with updated fields → save
      CARD_NOT_FOUND — ENTER with bad card number → error
      PF3_RETURN     — press PF3 → return to previous
    """
    scenario = inputs.get("SCENARIO", "FIRST_ENTRY")

    import copy
    repo = CardRepository(copy.deepcopy(_SEED_CARDS))

    commarea = CarddemoCommarea(
        cdemo_from_tranid="CCLI",
        cdemo_from_program="COCRDLIC",
        cdemo_user_id="USER0001",
        cdemo_user_type="U",
        cdemo_pgm_context=1,
    )

    if scenario == "FIRST_ENTRY":
        commarea.cdemo_pgm_context = 0
        result = process_card_update(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            card_input=CardUpdateInput(), card_repo=repo,
            preloaded_card_num="4111111111110001",
        )
    elif scenario == "LOOKUP_CARD":
        result = process_card_update(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            card_input=CardUpdateInput(card_num="4111111111110001"),
            card_repo=repo,
        )
    elif scenario == "UPDATE_CARD":
        result = process_card_update(
            eibcalen=100, eibaid="PF5", commarea=commarea,
            card_input=CardUpdateInput(
                card_num="4111111111110001", embossed_name="JANE UPDATED",
                expiration_date="2029-12-31", active_status="Y", cvv_cd="999",
            ),
            card_repo=repo,
        )
    elif scenario == "CARD_NOT_FOUND":
        result = process_card_update(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            card_input=CardUpdateInput(card_num="9999999999999999"),
            card_repo=repo,
        )
    elif scenario == "PF3_RETURN":
        result = process_card_update(
            eibcalen=100, eibaid="PF3", commarea=commarea,
            card_input=CardUpdateInput(), card_repo=repo,
        )
    else:
        commarea.cdemo_pgm_context = 0
        result = process_card_update(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            card_input=CardUpdateInput(), card_repo=repo,
            preloaded_card_num="4111111111110001",
        )

    card_num_out = ""
    embossed_name_out = ""
    if result.card_found:
        card_num_out = result.card_found.card_num
        embossed_name_out = result.card_found.card_embossed_name

    return {
        "ERROR": "Y" if result.error else "N",
        "SUCCESS": "Y" if result.success else "N",
        "MESSAGE": result.message,
        "XCTL_PROGRAM": result.xctl_program or "",
        "RETURN_TO_PREV": "Y" if result.return_to_prev else "N",
        "CARD_NUM": card_num_out,
        "EMBOSSED_NAME": embossed_name_out,
    }

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cocrdupc.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.util.*;

/**
 * Java reimplementation of COCRDUPC — CardDemo Credit Card Update Screen.
 *
 * <p>Mirrors {@code pipeline/reimpl/cocrdupc.py}. Python is the source of truth.
 *
 * <p>Standalone business method: {@link #processCardUpdate}. The {@link #runVector}
 * method is a thin adapter that builds scenario-specific parameters and delegates.
 */
public class Cocrdupc implements ProgramRunner {

    // ── Inner data types ─────────────────────────────────────────────────

    static class CardRecord {
        String cardNum;
        String acctId;
        String cvv;
        String embossedName;
        String expirationDate;
        String activeStatus;

        CardRecord(String cardNum, String acctId, String cvv,
                   String embossedName, String expirationDate, String activeStatus) {
            this.cardNum = cardNum;
            this.acctId = acctId;
            this.cvv = cvv;
            this.embossedName = embossedName;
            this.expirationDate = expirationDate;
            this.activeStatus = activeStatus;
        }

        CardRecord copy() {
            return new CardRecord(cardNum, acctId, cvv, embossedName, expirationDate, activeStatus);
        }
    }

    static class CardUpdateInput {
        String cardNum = "";
        String embossedName = "";
        String expirationDate = "";
        String activeStatus = "";
        String cvvCd = "";

        CardUpdateInput() {}

        CardUpdateInput(String cardNum, String embossedName, String expirationDate,
                        String activeStatus, String cvvCd) {
            this.cardNum = cardNum;
            this.embossedName = embossedName;
            this.expirationDate = expirationDate;
            this.activeStatus = activeStatus;
            this.cvvCd = cvvCd;
        }
    }

    static class CardUpdateResult {
        CardRecord cardFound = null;
        boolean error = false;
        boolean success = false;
        String message = "";
        String xctlProgram = "";
        boolean returnToPrev = false;
    }

    static class CardRepo {
        private final Map<String, CardRecord> cards;
        CardRepo(Map<String, CardRecord> cards) { this.cards = cards; }

        CardRecord find(String cardNum) {
            return cards.get(cardNum.trim());
        }

        boolean rewrite(CardRecord card) {
            String key = card.cardNum.trim();
            if (!cards.containsKey(key)) return false;
            cards.put(key, card);
            return true;
        }
    }

    // ── Standalone business method ──────────────────────────────────────

    /**
     * Process card update screen — mirrors Python's process_card_update.
     *
     * @param eibcalen         CICS EIBCALEN
     * @param eibaid           aid key pressed
     * @param pgmContext       commarea program context (0 = initial)
     * @param fromProgram      commarea cdemo_from_program
     * @param cardInput        update input fields
     * @param cardRepo         card repository
     * @param preloadedCardNum card number preloaded from previous screen
     * @return CardUpdateResult with all output fields
     */
    CardUpdateResult processCardUpdate(
            int eibcalen,
            String eibaid,
            int pgmContext,
            String fromProgram,
            CardUpdateInput cardInput,
            CardRepo cardRepo,
            String preloadedCardNum
    ) {
        CardUpdateResult result = new CardUpdateResult();

        if (eibcalen == 0) {
            result.returnToPrev = true;
            result.xctlProgram = "COSGN00C";
            return result;
        }

        if (pgmContext == 0) {
            if (preloadedCardNum != null && !preloadedCardNum.isEmpty()) {
                CardRecord card = cardRepo.find(preloadedCardNum);
                result.cardFound = card;
                if (card == null) {
                    result.error = true;
                    result.message = "Card " + preloadedCardNum.trim() + " not found";
                }
            }
            return result;
        }

        switch (eibaid) {
            case "ENTER":
                return lookupCard(cardInput.cardNum, cardRepo, result);
            case "PF3": {
                String back = (fromProgram != null && !fromProgram.isEmpty()) ? fromProgram : "COMEN01C";
                result.xctlProgram = back;
                result.returnToPrev = true;
                return result;
            }
            case "PF5":
                return saveCard(cardInput, cardRepo, result);
            case "PF12":
                result.xctlProgram = "COCRDLIC";
                result.returnToPrev = true;
                return result;
            default:
                result.error = true;
                result.message = "Invalid key pressed.";
                return result;
        }
    }

    private CardUpdateResult lookupCard(String cardNum, CardRepo repo, CardUpdateResult result) {
        if (cardNum == null || cardNum.trim().isEmpty()) {
            result.error = true;
            result.message = "Card number can NOT be empty...";
            return result;
        }
        CardRecord card = repo.find(cardNum);
        if (card == null) {
            result.error = true;
            result.message = "Card " + cardNum.trim() + " not found";
        } else {
            result.cardFound = card;
            result.message = "Press PF5 to save changes";
        }
        return result;
    }

    private CardUpdateResult saveCard(CardUpdateInput inp, CardRepo repo, CardUpdateResult result) {
        if (inp.cardNum == null || inp.cardNum.trim().isEmpty()) {
            result.error = true;
            result.message = "Card number can NOT be empty...";
            return result;
        }
        CardRecord card = repo.find(inp.cardNum);
        if (card == null) {
            result.error = true;
            result.message = "Card " + inp.cardNum.trim() + " not found";
            return result;
        }

        // Apply updates
        if (inp.embossedName != null && !inp.embossedName.trim().isEmpty()) {
            card.embossedName = inp.embossedName.length() > 50
                    ? inp.embossedName.substring(0, 50) : inp.embossedName;
        }
        if (inp.expirationDate != null && !inp.expirationDate.trim().isEmpty()) {
            card.expirationDate = inp.expirationDate.length() > 10
                    ? inp.expirationDate.substring(0, 10) : inp.expirationDate;
        }
        if (inp.activeStatus != null && !inp.activeStatus.trim().isEmpty()) {
            card.activeStatus = inp.activeStatus.trim().substring(0, 1);
        }
        if (inp.cvvCd != null && !inp.cvvCd.trim().isEmpty()) {
            try {
                Integer.parseInt(inp.cvvCd.trim());
                card.cvv = inp.cvvCd.trim();
            } catch (NumberFormatException e) {
                result.error = true;
                result.message = "CVV must be numeric";
                return result;
            }
        }

        boolean ok = repo.rewrite(card);
        if (ok) {
            result.success = true;
            result.cardFound = card;
            result.message = "Card " + card.cardNum.trim() + " has been updated ...";
        } else {
            result.error = true;
            result.message = "Unable to update card...";
        }
        return result;
    }

    // ── Seed data ───────────────────────────────────────────────────────

    private static final Map<String, CardRecord> SEED_CARDS = new LinkedHashMap<>();
    static {
        SEED_CARDS.put("4111111111110001", new CardRecord("4111111111110001", "10000001", "123", "JANE USER", "2028-12-31", "Y"));
        SEED_CARDS.put("4111111111110002", new CardRecord("4111111111110002", "10000001", "456", "BOB SMITH", "2027-06-30", "Y"));
    }

    private Map<String, CardRecord> deepCopySeedCards() {
        Map<String, CardRecord> copy = new LinkedHashMap<>();
        for (Map.Entry<String, CardRecord> e : SEED_CARDS.entrySet()) {
            copy.put(e.getKey(), e.getValue().copy());
        }
        return copy;
    }

    // ── Thin runVector adapter ──────────────────────────────────────────

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenario = inputs.getOrDefault("SCENARIO", "FIRST_ENTRY");

        CardRepo repo = new CardRepo(deepCopySeedCards());

        CardUpdateResult result;

        switch (scenario) {
            case "FIRST_ENTRY":
                result = processCardUpdate(100, "ENTER", 0, "COCRDLIC",
                        new CardUpdateInput(), repo, "4111111111110001");
                break;

            case "LOOKUP_CARD":
                result = processCardUpdate(100, "ENTER", 1, "COCRDLIC",
                        new CardUpdateInput("4111111111110001", "", "", "", ""),
                        repo, "");
                break;

            case "UPDATE_CARD":
                result = processCardUpdate(100, "PF5", 1, "COCRDLIC",
                        new CardUpdateInput("4111111111110001", "JANE UPDATED", "2029-12-31", "Y", "999"),
                        repo, "");
                break;

            case "CARD_NOT_FOUND":
                result = processCardUpdate(100, "ENTER", 1, "COCRDLIC",
                        new CardUpdateInput("9999999999999999", "", "", "", ""),
                        repo, "");
                break;

            case "PF3_RETURN":
                result = processCardUpdate(100, "PF3", 1, "COCRDLIC",
                        new CardUpdateInput(), repo, "");
                break;

            default:
                result = processCardUpdate(100, "ENTER", 0, "COCRDLIC",
                        new CardUpdateInput(), repo, "4111111111110001");
                break;
        }

        String cardNumOut = "";
        String embossedNameOut = "";
        if (result.cardFound != null) {
            cardNumOut = result.cardFound.cardNum;
            embossedNameOut = result.cardFound.embossedName;
        }

        Map<String, String> out = new LinkedHashMap<>();
        out.put("ERROR", result.error ? "Y" : "N");
        out.put("SUCCESS", result.success ? "Y" : "N");
        out.put("MESSAGE", result.message);
        out.put("XCTL_PROGRAM", result.xctlProgram);
        out.put("RETURN_TO_PREV", result.returnToPrev ? "Y" : "N");
        out.put("CARD_NUM", cardNumOut);
        out.put("EMBOSSED_NAME", embossedNameOut);
        return out;
    }
}
